kafka-client/src/main/java/dev/responsive/kafka/internal/db/WriterFactory.java

```diff
@@ -16,13 +16,103 @@
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ExecutionException;
+import org.apache.kafka.common.utils.LogContext;
+import org.slf4j.Logger;
 
-public interface WriterFactory<K> {
+public abstract class WriterFactory<K, P> {
+  private final Logger log;
+  private final String logPrefix;
 
-  RemoteWriter<K> createWriter(
-      final SessionClients client,
-      final int partition
+  public WriterFactory(final String logPrefix) {
+    this.logPrefix = logPrefix;
+    this.log = new LogContext(logPrefix).logger(WriterFactory.class);
+  }
+
+  public abstract RemoteWriter<K, P> createWriter(
+      final P tablePartition
+  );
+
+  public abstract String tableName();
+
+  protected abstract P tablePartitionForKey(
+      final K key
   );
 
+  protected abstract RemoteWriteResult<P> setOffset(final long consumedOffset);
+
+  protected abstract long offset();
+
+  public PendingFlush beginNewFlush() {
+    return new PendingFlush();
+  }
+
+  public RemoteWriteResult<P> commitPendingFlush(
+      final PendingFlush pendingFlush,
+      final long consumedOffset
+  ) {
+    return pendingFlush.completeFlush(consumedOffset);
+  }
+
+  public String failedFlushError(
+      final RemoteWriteResult<P> result,
+      final long consumedOffset
+  ) {
+    final long offset = offset();
+    return String.format(
+        "%sError while writing batch for table partition %s! Batch Offset: %d, Persisted Offset: %d",
+        logPrefix,
+        result.tablePartition(),
+        consumedOffset,
+        offset
+    );
+  }
+
+  public class PendingFlush {
+    final Map<P, RemoteWriter<K, P>> activeWriters = new HashMap<>();
+
+    public int numRemoteWriters() {
+      return activeWriters.size();
+    }
+
+    public RemoteWriter<K, P> writerForKey(
+        final K key
+    ) {
+      final P tablePartition = tablePartitionForKey(key);
+      return activeWriters.computeIfAbsent(
+          tablePartition,
+          WriterFactory.this::createWriter
+      );
+    }
+
+    private RemoteWriteResult<P> completeFlush(final long consumedOffset) {
+      final P ignored = null;
+      // initialize with null as the first partial result will always fail the !isApplied check and
+      // be discarded in favor of the next, and real, WriteResult in the loop below
+      CompletionStage<RemoteWriteResult<P>> partialResult =
+          CompletableFuture.completedStage(RemoteWriteResult.success(ignored));
+      for (final RemoteWriter<K, P> writer : activeWriters.values()) {
+        partialResult = partialResult.thenCombine(writer.flush(), (one, two) -> !one.wasApplied() ? one : two);
+      }
+
+      try {
+        final RemoteWriteResult<P> result = partialResult.toCompletableFuture().get();
+        log.debug("Succesfully flushed writes to table partition {}", result.tablePartition());
```

Comment:
do we need to check if the result was applied before logging this?

Reply:
good catch